cvmfs/loader_talk.cc

```diff
@@ -136,19 +136,58 @@ int MainReload(const std::string &socket_path, const bool stop_and_go,
   }
   LogCvmfs(kLogCvmfs, kLogStdout, "done");
 
-
   // reload mode: debug (d) or non-debug (n)
-  char command = debug ? 'd' : 'n';
-  WritePipe(socket_fd, &command, 1);
-
-  command = stop_and_go ? 'S' : 'R';
-  WritePipe(socket_fd, &command, 1);
+  char commands[2];
+  commands[0] = debug ? 'd' : 'n';
+  commands[1] = stop_and_go ? 'S' : 'R';
+
+  // Loaders before version 2.11 won't recognize 'd' or 'n'. They will
+  // send "unknown command" and close the connection. We try to send
+  // both the commands and react according to what we read from the socket.
+
+  ssize_t retval;
+  do {
+    retval = send(socket_fd, commands, 2, MSG_NOSIGNAL);
+  } while ((retval <= 0) && (errno == EINTR));
+
+  if (retval <= 0) {
+    LogCvmfs(kLogCvmfs, kLogStderr, "Sending reload command failed!");
+    return 103;
+  }
 
   char buf;
-  int retval;
+  std::string first_line;
+  bool past_first_line = false;
   while ((retval = read(socket_fd, &buf, 1)) == 1) {
     if (buf == '~')
       break;
+
+    if (first_line == "unknown command") {
+      // We have a pre-2.11 loader, reconnect to the socket
+      LogCvmfs(kLogCvmfs, kLogStdout,
+               "Connecting in backwards compatibility mode");
+      close(socket_fd);
+      socket_fd = ConnectSocket(socket_path);
+      if (socket_fd < 0) {
+        LogCvmfs(kLogCvmfs, kLogStderr, "reconnecting failed!");
+        return 104;
+      }
+      WritePipe(socket_fd, &commands[1], 1);
+      first_line.clear();
+      past_first_line = true;
+      continue;
+    }
+
+    if (!past_first_line) {
+      if (buf == '\n') {
+        LogCvmfs(kLogCvmfs, kLogStdout, "%s", first_line.c_str());
```

Comment:
if we print it here: i think we will print double  `unkown command unkown command\n\n`  due to the logcvmfs in line 191 (also note because of this if here we will print the full command again before printing another `\n`

Reply:
I don't think so: if we're not past the first line, there is the `continue` on line 188, and if we are past the first line this log won't be executed - but let me double check.

Reply:
on the second look i think you are right..
but then i suggest deleteing the logcvmfs in line 191 and the continue statements in 178 and 188. they seem unnecessary.

Reply:
It's necessary: for printing everything that comes over the socket after "Connecting in backwards compatibility mode"
```
alice.cern.ch: Debug: using library /usr/lib/libcvmfs_fuse3_stub.so
alice.cern.ch: Connecting to CernVM-FS loader... done
alice.cern.ch: Connecting in backwards compatibility mode
alice.cern.ch: Entering maintenance mode
alice.cern.ch: Draining out kernel caches (up to 60s)
alice.cern.ch: Blocking new file system calls
alice.cern.ch: Waiting for active file system calls
alice.cern.ch: Saving inode tracker
alice.cern.ch: Saving negative entry cache
alice.cern.ch: Saving page cache entry tracker
alice.cern.ch: Saving chunk tables
alice.cern.ch: Saving inode generation
alice.cern.ch: Saving fuse state
alice.cern.ch: Saving open files table
alice.cern.ch: Unloading Fuse module
alice.cern.ch: Waiting for the delivery of SIGUSR1...
alice.cern.ch: Re-Loading Fuse module
alice.cern.ch: Restoring inode tracker...  done
alice.cern.ch: Restoring dentry tracker...  done
alice.cern.ch: Restoring page cache entry tracker...  done
alice.cern.ch: Restoring chunk tables...  done
alice.cern.ch: Restoring inode generation...  done
alice.cern.ch: Restoring fuse state...  done
alice.cern.ch: Restoring open files table... done
alice.cern.ch: Restoring open files counter...  done
alice.cern.ch: Releasing saved glue buffer
alice.cern.ch: Releasing saved dentry tracker
alice.cern.ch: Releasing saved page cache entry cache
alice.cern.ch: Releasing chunk tables
alice.cern.ch: Releasing saved inode generation info
alice.cern.ch: Releasing fuse state
alice.cern.ch: Releasing saved open files table
alice.cern.ch: Releasing open files counter
alice.cern.ch: Activating Fuse module
```